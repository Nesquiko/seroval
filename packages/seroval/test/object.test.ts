import { describe, it, expect } from 'vitest';
import {
  AsyncServerValue,
  deserialize,
  fromJSON,
  serialize,
  serializeAsync,
  ServerValue,
  toJSON,
  toJSONAsync,
} from '../src';

describe('objects', () => {
  describe('serialize', () => {
    it('supports Objects', () => {
      const example = { hello: 'world' };
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Record<string, string>>(result);
      expect(back).toBeInstanceOf(Object);
      expect(back.hello).toBe(example.hello);
    });
    it('supports self-recursion', () => {
      const example: Record<string, ServerValue> = {};
      example.a = example;
      example.b = example;
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Record<string, ServerValue>>(result);
      expect(back.a).toBe(back);
      expect(back.b).toBe(back);
    });
  });
  describe('serializeAsync', () => {
    it('supports Objects', async () => {
      const example = { hello: 'world' };
      const result = await serializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<Record<string, string>>>(result);
      expect(back).toBeInstanceOf(Object);
      expect(back.hello).toBe(example.hello);
    });
    it('supports self-recursion', async () => {
      const example: Record<string, AsyncServerValue> = {};
      example.a = Promise.resolve(example);
      example.b = Promise.resolve(example);
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Record<string, Promise<any>>>(result);
      expect(await back.a).toBe(back);
      expect(await back.b).toBe(back);
    });
  });
  describe('toJSON', () => {
    it('supports Objects', () => {
      const example = { hello: 'world' };
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Record<string, string>>(result);
      expect(back).toBeInstanceOf(Object);
      expect(back.hello).toBe(example.hello);
    });
    it('supports self-recursion', () => {
      const example: Record<string, ServerValue> = {};
      example.a = example;
      example.b = example;
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Record<string, ServerValue>>(result);
      expect(back.a).toBe(back);
      expect(back.b).toBe(back);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Objects', async () => {
      const example = { hello: 'world' };
      const result = await toJSONAsync(Promise.resolve(example));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<Record<string, string>>>(result);
      expect(back).toBeInstanceOf(Object);
      expect(back.hello).toBe(example.hello);
    });
    it('supports self-recursion', async () => {
      const example: Record<string, AsyncServerValue> = {};
      example.a = Promise.resolve(example);
      example.b = Promise.resolve(example);
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Record<string, Promise<any>>>(result);
      expect(await back.a).toBe(back);
      expect(await back.b).toBe(back);
    });
  });
});
