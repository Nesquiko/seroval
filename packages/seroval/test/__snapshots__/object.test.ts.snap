// Vitest Snapshot v1

exports[`objects > serialize > supports Objects 1`] = `"({hello:\\"world\\"})"`;

exports[`objects > serialize > supports self-recursion 1`] = `"(h=>(h={},h.b=h.a=h,h))()"`;

exports[`objects > serializeAsync > supports Objects 1`] = `"Promise.resolve({hello:\\"world\\"})"`;

exports[`objects > serializeAsync > supports self-recursion 1`] = `"(h=>(h={a:Promise.resolve().then(()=>h),b:Promise.resolve().then(()=>h)}))()"`;

exports[`objects > toJSON > supports Objects 1`] = `"{\\"t\\":{\\"t\\":16,\\"i\\":0,\\"d\\":{\\"k\\":[\\"hello\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"world\\"}],\\"s\\":1}},\\"r\\":0,\\"i\\":true,\\"f\\":8191,\\"m\\":[]}"`;

exports[`objects > toJSON > supports self-recursion 1`] = `"{\\"t\\":{\\"t\\":16,\\"i\\":0,\\"d\\":{\\"k\\":[\\"a\\",\\"b\\"],\\"v\\":[{\\"t\\":10,\\"i\\":0},{\\"t\\":10,\\"i\\":0}],\\"s\\":2}},\\"r\\":0,\\"i\\":true,\\"f\\":8191,\\"m\\":[0]}"`;

exports[`objects > toJSONAsync > supports Objects 1`] = `"{\\"t\\":{\\"t\\":18,\\"i\\":0,\\"f\\":{\\"t\\":16,\\"i\\":1,\\"d\\":{\\"k\\":[\\"hello\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"world\\"}],\\"s\\":1}}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`objects > toJSONAsync > supports self-recursion 1`] = `"{\\"t\\":{\\"t\\":16,\\"i\\":0,\\"d\\":{\\"k\\":[\\"a\\",\\"b\\"],\\"v\\":[{\\"t\\":18,\\"i\\":1,\\"f\\":{\\"t\\":10,\\"i\\":0}},{\\"t\\":18,\\"i\\":2,\\"f\\":{\\"t\\":10,\\"i\\":0}}],\\"s\\":2}},\\"r\\":0,\\"i\\":true,\\"f\\":8191,\\"m\\":[0]}"`;
